src/Orleans/Messaging/ProxiedMessageCenter.cs

```diff
@@ -287,6 +293,12 @@ public Message WaitMessage(Message.Categories type, CancellationToken ct)
         {
             try
             {
+                if (ct.IsCancellationRequested)
+                {
+                    return null;
+                }
+
+                // Don't pass CancellationToken to Take. It causes too much spinning.
                 Message msg = PendingInboundMessages.Take();
```

Comment:
I think I will just remove the Dispose completely. Its redundant. I should not have put it. CompleteAdding is just enough. No need to "smash kill with a hammer" if a clean stop does the work. 
If we did want Dispose (which I think we do not now), then we need to wait for this thread to finish first and dispose only after we know there are no usages.